internal/git/util.go

```diff
@@ -245,14 +268,177 @@ func cloneRepo(ctx context.Context, gitSync *v1alpha1.GitSync, options *git.Clon
 	path := getLocalRepoPath(gitSync)
 
 	r, err := git.PlainCloneContext(ctx, path, false, options)
-	if err != nil && errors.Is(err, git.ErrRepositoryAlreadyExists) {
-		// open the existing repo and return it.
-		existingRepo, openErr := git.PlainOpen(path)
-		if openErr != nil {
-			return r, fmt.Errorf("failed to open existing repo, err: %v", openErr)
+	if err != nil {
+		if errors.Is(err, git.ErrRepositoryAlreadyExists) {
+			// Open the existing repo and return it.
+			existingRepo, openErr := git.PlainOpen(path)
+			if openErr != nil {
+				return nil, fmt.Errorf("failed to open existing repo, err: %v", openErr)
+			}
+			return existingRepo, nil
+		}
+		return nil, err
+	}
+
+	// No need to fetch the references if  the target revision is already main or master (branches)
+	if gitSync.Spec.TargetRevision != "main" && gitSync.Spec.TargetRevision != "master" {
+		// fetch all references
+		err = fetchAll(r)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Perform checkout to the specified reference after a successful clone
+	if checkoutErr := checkoutRepo(r, gitSync.Spec.TargetRevision); checkoutErr != nil {
```

Comment:
Is it required to run in each reconsile? what if already checkedout?

Reply:
yeah that can be checked with current branch == targetRevsion.

Reply:
As per discussion, it's running for initial clone only, so it looks fine then.

Reply:
I believe this code will be run each time when we reconcile. so we should avoid checking out if already does each time, right @shubhamdixit863 ?

Reply:
@xdevxy in this code above -```
r, err := git.PlainCloneContext(ctx, path, false, options)
	if err != nil {
		if errors.Is(err, git.ErrRepositoryAlreadyExists) {
			// Open the existing repo and return it.
			existingRepo, openErr := git.PlainOpen(path)
			if openErr != nil {
				return nil, fmt.Errorf("failed to open existing repo, err: %v", openErr)
			}
			return existingRepo, nil
		}
		return nil, err
	}
```
we are already checking if the repository is cloned or not  `if errors.Is(err, git.ErrRepositoryAlreadyExists) {` so if the repository is already cloned ,it means we have already checked it out ,so i guess it doesn't runs multiple times during reconcilation